backend/src/api/fiat-conversion.ts

```diff
@@ -33,48 +35,75 @@ class FiatConversion {
   }
 
   private async updateCurrency(): Promise<void> {
-    const headers = { 'User-Agent': `mempool/v${backendInfo.getBackendInfo().version}` };
-    let fiatConversionUrl: string;
-    let response: AxiosResponse;
-
-    try {
-      if (config.SOCKS5PROXY.ENABLED) {
-        let socksOptions: any = {
-          agentOptions: {
-            keepAlive: true,
-          },
-          hostname: config.SOCKS5PROXY.HOST,
-          port: config.SOCKS5PROXY.PORT
-        };
+    type axiosOptions = {
+      headers: {
+        'User-Agent': string
+      };
+      timeout: number;
+      httpAgent?: http.Agent;
+      httpsAgent?: https.Agent;
+    }
+    const setDelay = (secs: number = 1): Promise<void> => new Promise(resolve => setTimeout(() => resolve(), secs * 1000));
+    const fiatConversionUrl = (config.SOCKS5PROXY.ENABLED === true) && (config.SOCKS5PROXY.USE_ONION === true) ? config.PRICE_DATA_SERVER.TOR_URL : config.PRICE_DATA_SERVER.CLEARNET_URL;
+    const isHTTP = (new URL(fiatConversionUrl).protocol.split(':')[0] === 'http') ? true : false;
+    const axiosOptions: axiosOptions = {
+      headers: {
+        'User-Agent': (config.MEMPOOL.USER_AGENT === 'mempool') ? `mempool/v${backendInfo.getBackendInfo().version}` : `${config.MEMPOOL.USER_AGENT}`
+      },
+      timeout: config.SOCKS5PROXY.ENABLED ? 30000 : 10000
+    };
 
-        if (config.SOCKS5PROXY.USERNAME && config.SOCKS5PROXY.PASSWORD) {
-          socksOptions.username = config.SOCKS5PROXY.USERNAME;
-          socksOptions.password = config.SOCKS5PROXY.PASSWORD;
-        }
+    let retry = 0;
 
-        const agent = new SocksProxyAgent(socksOptions);
-        fiatConversionUrl = config.PRICE_DATA_SERVER.TOR_URL;
-        logger.debug('Querying currency rates service...');
-        response = await axios.get(fiatConversionUrl, { httpAgent: agent, headers: headers, timeout: 30000 });
-      } else {
-        fiatConversionUrl = config.PRICE_DATA_SERVER.CLEARNET_URL;
+    while(retry < config.MEMPOOL.EXTERNAL_MAX_RETRY) {
```

Comment:
@ayanamidev I missed this on the previous PR but the retries are blocking the backend from performing other tasks. Notice the logs after getting 403s below

```
May 23 21:37:58 [99503] NOTICE: Mempool Server is running on port 8999
May 23 21:37:58 [99503] INFO: Updating latest mining pools from Github
May 23 21:37:58 [99503] INFO: List of public pools will be queried over the Tor network
May 23 21:37:58 [99503] ERR: Cannot fetch pools.json sha from db. Reason: connect ENOENT /var/run/mysql/mysql.sock
May 23 21:37:59 [99503] ERR: Could not connect to Github. Reason: Request failed with status code 403
May 23 21:37:59 [99503] DEBUG: USD Conversion Rate: 29306.538
May 23 21:39:00 [99503] ERR: Could not connect to Github. Reason: Request failed with status code 403
May 23 21:40:01 [99503] ERR: Could not connect to Github. Reason: Request failed with status code 403
May 23 21:41:03 [99503] ERR: Could not connect to Github. Reason: Request failed with status code 403
May 23 21:42:04 [99503] ERR: Could not connect to Github. Reason: Request failed with status code 403
May 23 21:43:05 [99503] DEBUG: Pools.json sha | Current: undefined | Github: cc8d3e01dea2787525ed1f182a0f9a685fe3a850
May 23 21:43:05 [99503] WARN: Pools.json is outdated, fetch latest from github
May 23 21:43:07 [99503] DEBUG: Parse coinbase_tags
May 23 21:43:07 [99503] DEBUG: Parse payout_addresses
May 23 21:43:07 [99503] DEBUG: Identify unique mining pools
May 23 21:43:07 [99503] DEBUG: Found 136 unique mining pools
May 23 21:43:07 [99503] ERR: Cannot get existing pools from the database, skipping pools.json import
May 23 21:43:07 [99503] ERR: Cannot save github pools.json sha into the db. Reason: connect ENOENT /var/run/mysql/mysql.sock
May 23 21:43:07 [99503] NOTICE: PoolsUpdater completed
May 23 21:43:07 [99503] DEBUG: Initial difficulty adjustment data set.
May 23 21:43:07 [99503] DEBUG: New block found (#737663)!
May 23 21:43:07 [99503] DEBUG: Indexing tx 1 of 830 in block #737663
May 23 21:43:07 [99503] DEBUG: 0 of 830 found in mempool. 1 fetched through backend service.
```

or when we can't connect to the proxy:

```
May 23 22:00:04 [102230] INFO: Updating latest mining pools from Github
May 23 22:00:04 [102230] INFO: List of public pools will be queried over the Tor network
May 23 22:00:04 [102230] ERR: Cannot fetch pools.json sha from db. Reason: connect ENOENT /var/run/mysql/mysql.sock
May 23 22:00:04 [102230] ERR: Could not connect to Github. Reason: connect ECONNREFUSED 127.0.0.1:9051
retry: 1
May 23 22:01:04 [102230] DEBUG: Querying currency rates service...
May 23 22:01:04 [102230] ERR: Error updating fiat conversion rates: connect ECONNREFUSED 127.0.0.1:9051
May 23 22:01:04 [102230] ERR: Could not connect to Github. Reason: connect ECONNREFUSED 127.0.0.1:9051
retry: 2
May 23 22:02:04 [102230] DEBUG: Querying currency rates service...
May 23 22:02:04 [102230] ERR: Error updating fiat conversion rates: connect ECONNREFUSED 127.0.0.1:9051
May 23 22:02:04 [102230] ERR: Could not connect to Github. Reason: connect ECONNREFUSED 127.0.0.1:9051
retry: 3

``` 

```

Reply:
This seems odd, maybe the code isn't asynchronous and would run with await keyword instead which is not very efficient as well

Reply:
@knorrium I believe you should ask @nymkappa about the issue since the reason for blocking tasks is mainly the pool updater has 'await' keyword to wait until Promise resolves https://github.com/mempool/mempool/blame/master/backend/src/index.ts#L155.

And since the indexer function https://github.com/mempool/mempool/blame/master/backend/src/index.ts#L158 depends on https://github.com/mempool/mempool/blame/master/backend/src/index.ts#L155, removing the await keyword would likely break other updater's behavior.

So my recommendation is to go as is but if people want, they would be able to reduce the retry limit per job.

Reply:
Have addressed issues with Tor connections https://github.com/mempool/mempool/pull/1691/commits/a1fbe582b4724fdb43b6208ac8d6dc852d8a3893, maybe try this commit.